engine/access/rest/routes/subscribe_events_test.go

```diff
@@ -0,0 +1,338 @@
+package routes
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+	"testing"
+	"time"
+
+	"golang.org/x/exp/slices"
+
+	mocks "github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/onflow/flow-go/engine/access/rest/request"
+	"github.com/onflow/flow-go/engine/access/state_stream"
+	mockstatestream "github.com/onflow/flow-go/engine/access/state_stream/mock"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+type testType struct {
+	name         string
+	startBlockID flow.Identifier
+	startHeight  uint64
+
+	eventTypes []string
+	addresses  []string
+	contracts  []string
+}
+
+var testEventTypes = []flow.EventType{
+	"A.0123456789abcdef.flow.event",
+	"B.0123456789abcdef.flow.event",
+	"C.0123456789abcdef.flow.event",
+}
+
+type SubscribeEventsSuite struct {
+	suite.Suite
+
+	blocks      []*flow.Block
+	blockEvents map[flow.Identifier]flow.EventsList
+}
+
+func TestSubscribeEventsSuite(t *testing.T) {
+	suite.Run(t, new(SubscribeEventsSuite))
+}
+
+func (s *SubscribeEventsSuite) SetupTest() {
+	rootBlock := unittest.BlockFixture()
+	parent := rootBlock.Header
+
+	blockCount := 5
+
+	s.blocks = make([]*flow.Block, 0, blockCount)
+	s.blockEvents = make(map[flow.Identifier]flow.EventsList, blockCount)
+
+	for i := 0; i < blockCount; i++ {
+		block := unittest.BlockWithParentFixture(parent)
+		// update for next iteration
+		parent = block.Header
+
+		result := unittest.ExecutionResultFixture()
+		blockEvents := unittest.BlockEventsFixture(block.Header, (i%len(testEventTypes))*3+1, testEventTypes...)
+
+		s.blocks = append(s.blocks, block)
+		s.blockEvents[block.ID()] = blockEvents.Events
+
+		s.T().Logf("adding exec data for block %d %d %v => %v", i, block.Header.Height, block.ID(), result.ExecutionDataID)
+	}
+}
+
+func (s *SubscribeEventsSuite) TestSubscribeEvents() {
```

Comment:
A short godoc for each test would be appreciated, just to get general idea of what we are dealing with